src/envoy/http/service_control/handler_impl_test.cc

```diff
@@ -1063,6 +1065,80 @@ TEST_F(HandlerTest, HandlerReportWithoutCheck) {
   handler.callReport(&headers, &response_headers, &resp_trailer_);
 }
 
+class HandlerReportStatusTest : public HandlerTest {
+ protected:
+  void runTest(unsigned int http_response_code,
+               TestResponseHeaderMapImpl response_headers,
+               TestResponseTrailerMapImpl response_trailers,
+               Code expected_grpc_status) {
+    EXPECT_CALL(mock_stream_info_, responseCode())
+        .WillRepeatedly(Return(http_response_code));
+
+    setPerRouteOperation("get_header_key");
+    TestRequestHeaderMapImpl headers{
+        {":method", "GET"}, {":path", "/echo"}, {"x-api-key", "foobar"}};
+    CheckDoneFunc stored_on_done;
+    CheckResponseInfo response_info;
+    ServiceControlHandlerImpl handler(headers, mock_stream_info_, "test-uuid",
+                                      *cfg_parser_, test_time_, stats_);
+
+    ReportRequestInfo expected_report_info;
+    initExpectedReportInfo(expected_report_info);
+    expected_report_info.api_key = "foobar";
+    expected_report_info.status = Status::OK;
+    expected_report_info.http_response_code = http_response_code;
+    expected_report_info.grpc_response_code = expected_grpc_status;
+    EXPECT_CALL(*mock_call_, callReport(MatchesReportInfo(
+                                 expected_report_info, headers,
+                                 response_headers, response_trailers)));
+    handler.callReport(&headers, &response_headers, &response_trailers);
+  }
+};
+
+TEST_F(HandlerReportStatusTest, GrpcStatusFromTrailers) {
+  // Test: When grpc status trailer is present, it's value will be used in
+  // addition to HTTP status.
+  runTest(200, {{"content-type", "application/grpc"}}, {{"grpc-status", "7"}},
+          Code::PERMISSION_DENIED);
+}
+
+TEST_F(HandlerReportStatusTest, GrpcStatusFromHeaders) {
+  // Test: When grpc status header is present, it's value will be used in
+  // addition to HTTP status.
+  // Remember: grpc-status will show up in the headers during a trailers-only
+  // response.
+  runTest(200, {{"content-type", "application/grpc"}}, {{"grpc-status", "14"}},
+          Code::UNAVAILABLE);
+}
+
+TEST_F(HandlerReportStatusTest, MissingGrpcStatus) {
+  // Test: When grpc status header/trailer is missing, we fallback to HTTP
+  // status.
+  runTest(200, {{"content-type", "application/grpc"}}, {}, Code::OK);
```

Comment:
If there is not grpc-status, I prefer not to write "grpc_status_code" into log entry.   Now, we are still writting OK,  is that rigth?

Reply:
Correct, fixed now. Use `absl::optional` based on guidance from go/totw/123. This should also reduce some duplicate logic in the request builder.

Also added an integration test for non-canonical rpc code.